botocore/configprovider.py

```diff
@@ -851,3 +858,149 @@ def provide(self):
+
+        :type client_name: str
+        :param client_name: client name.
+
+        :type environ: dict
+        :param environ: A mapping to use for environment variables. If this
+            is not provided it will default to use os.environ.
+        """
+        self._full_config = full_config
+        self._scoped_config = scoped_config
+        self._client_name = client_name
+
+        if environ is None:
+            environ = os.environ
+        self._environ = environ
+
+    def provide(self):
+        """Lookup the configured endpoint URL.
+
+        The order is:
+
+        1. The value provided by a service-specific environment variable.
+        2. The value provided by the global endpoint environment variable
+           (AWS_ENDPOINT_URL).
+        3. The value provided by a service-specific parameter from a services
+           definition section in the shared configuration file.
+        4. The value provided by the global parameter from a services
+           definition section in the shared configuration file.
+        """
+        for location in self._ENDPOINT_URL_LOOKUP_ORDER:
+            logger.debug(
+                'Looking for endpoint for %s via: %s',
+                self._client_name,
+                location,
+            )
+
+            endpoint_url = getattr(self, f'_get_endpoint_url_{location}')()
+
+            if endpoint_url:
+                logger.info(
+                    'Found endpoint for %s via: %s.',
+                    self._client_name,
+                    location,
+                )
+                return endpoint_url
+
+        logger.debug('No configured endpoint found.')
+        return None
+
+    def unhyphenize_service_name(self, service_name):
+        return service_name.replace('-', ' ')
+
+    def _get_service_id(self, client_name):
+        client_name = utils.handle_service_name_alias(client_name)
+
+        client_name = (
+            utils._CLIENT_NAME_TO_HYPHENIZED_SERVICE_ID_OVERRIDES.get(
+                client_name, client_name
+            )
+        )
+
+        return self.unhyphenize_service_name(client_name)
+
+    def _snakecase_service_id(self, service_id):
+        return service_id.replace(' ', '_')
+
+    def _get_service_env_var_name(self):
+        transformed_service_id_env = self._snakecase_service_id(
+            self._get_service_id(self._client_name)
+        ).upper()
+        return f'AWS_ENDPOINT_URL_{transformed_service_id_env}'
+
+    def _get_services_config(self):
+        if 'services' not in self._scoped_config:
+            return {}
+
+        section_name = self._scoped_config['services']
+        services_section = self._full_config.get('services', {}).get(
+            section_name
+        )
+
+        if not services_section:
+            error_msg = (
+                f'The profile is configured to use the services '
+                f'section but the "{section_name}" services '
+                f'configuration does not exist.'
```

Comment:
```suggestion
                'configuration does not exist.'
```

Reply:
I made changes like this previously and it was suggested that for uniformity (and clarity that its an f-string) to use the `f` syntax on all string sections.

Reply:
When was that suggested? I've never heard of doing that before. If you've gotten that feedback, keep it, but I find it weird the need to clarify an f string on every line when its abundantly obvious where the formatting is happening, particularly in modern IDEs with good syntax highlighting.